ktor-shared/ktor-serialization/ktor-serialization-kotlinx/common/src/io/ktor/shared/serialization/kotlinx/KotlinxWebsocketSerializationConverter.kt

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2014-2021 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.shared.serialization.kotlinx
+
+import io.ktor.http.*
+import io.ktor.http.cio.websocket.*
+import io.ktor.http.content.*
+import io.ktor.serialization.*
+import io.ktor.serialization.kotlinx.*
+import io.ktor.util.reflect.*
+import io.ktor.utils.io.*
+import io.ktor.utils.io.charsets.*
+import io.ktor.utils.io.core.*
+import kotlinx.serialization.*
+
+/**
+ * Creates a converter for websocket serializing with the specified string [format] and
+ * [defaultCharset] (optional, usually it is UTF-8).
+ */
+@OptIn(ExperimentalSerializationApi::class)
+public class KotlinxWebsocketSerializationConverter(
+    private val format: SerialFormat,
+) : WebsocketContentConverter {
+
+    init {
+        require(format is BinaryFormat || format is StringFormat) {
+            "Only binary and string formats are supported, " +
+                "$format is not supported."
+        }
+    }
+
+    override suspend fun serialize(charset: Charset, typeInfo: TypeInfo, value: Any): Frame {
+        val result = try {
```

Comment:
can you remove code duplication here?

Reply:
Moved serialization implementation to `KotlinxSerializationBase`